File: Kalorie/Kalorie/Features/Dashboard/FoodConsumedDetailConfigurator.swift
//
//  FoodConsumedDetailConfigurator.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.07.2026.
//

import Foundation

struct FoodConsumedDetailConfigurator {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func createView(food: FoodConsumedDomain, mealTypes: [MealTypeDomain], onFoodUpdated: @escaping () -> Void) -> FoodConsumedDetailView {
        FoodConsumedDetailView(
            viewModel: FoodConsumedDetailViewModel(
                food: food,
                mealTypes: mealTypes,
                updateFoodConsumed: UpdateFoodConsumedUseCase(dataProvider: dataProvider, authProvider: authProvider),
                assignFoodMealType: AssignFoodMealTypeUseCase(dataProvider: dataProvider, authProvider: authProvider),
                fetchMealTypes: FetchMealTypesUseCase(dataProvider: dataProvider, authProvider: authProvider),
                isFavouriteFood: IsFavouriteFoodUseCase(dataProvider: dataProvider, authProvider: authProvider),
                addFavouriteFood: AddFavouriteFoodUseCase(dataProvider: dataProvider, authProvider: authProvider),
                removeFavouriteFood: RemoveFavouriteFoodUseCase(dataProvider: dataProvider, authProvider: authProvider),
                fetchFoodItemByBarcode: FetchFoodItemByBarcodeUseCase(dataProvider: dataProvider),
                fetchFoodByBarcodeExternally: FetchFoodByBarcodeExternallyUseCase(),
                onFoodUpdated: onFoodUpdated
            )
        )
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/dashboard/FoodConsumedDetailConfigurator.kt
package antoni.kalorie.features.dashboard

import androidx.compose.runtime.Composable
import androidx.lifecycle.viewmodel.compose.viewModel
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodConsumedDomain
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.usecases.AddFavouriteFoodUseCase
import antoni.kalorie.core.usecases.AssignFoodMealTypeUseCase
import antoni.kalorie.core.usecases.FetchFoodByBarcodeExternallyUseCase
import antoni.kalorie.core.usecases.FetchFoodItemByBarcodeUseCase
import antoni.kalorie.core.usecases.FetchMealTypesUseCase
import antoni.kalorie.core.usecases.FetchMyFoodItemReportUseCase
import antoni.kalorie.core.usecases.IsFavouriteFoodUseCase
import antoni.kalorie.core.usecases.RemoveFavouriteFoodUseCase
import antoni.kalorie.core.usecases.SubmitFoodItemReportUseCase
import antoni.kalorie.core.usecases.UpdateFoodConsumedUseCase

class FoodConsumedDetailConfigurator(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) {

    // MARK: - Functions

    @Composable
    fun createView(
        food: FoodConsumedDomain,
        mealTypes: List<MealTypeDomain>,
        onBack: () -> Unit,
        onFoodUpdated: () -> Unit,
    ) {
        val viewModel = viewModel {
            FoodConsumedDetailViewModel(
                food = food,
                mealTypes = mealTypes,
                updateFoodConsumed = UpdateFoodConsumedUseCase(dataProvider, authProvider),
                assignFoodMealType = AssignFoodMealTypeUseCase(dataProvider, authProvider),
                fetchMealTypes = FetchMealTypesUseCase(dataProvider, authProvider),
                isFavouriteFood = IsFavouriteFoodUseCase(dataProvider, authProvider),
                addFavouriteFood = AddFavouriteFoodUseCase(dataProvider, authProvider),
                removeFavouriteFood = RemoveFavouriteFoodUseCase(dataProvider, authProvider),
                fetchFoodItemByBarcode = FetchFoodItemByBarcodeUseCase(dataProvider),
                fetchFoodByBarcodeExternally = FetchFoodByBarcodeExternallyUseCase(),
                fetchMyFoodItemReport = FetchMyFoodItemReportUseCase(dataProvider, authProvider),
                submitFoodItemReport = SubmitFoodItemReportUseCase(dataProvider, authProvider),
                onFoodUpdated = onFoodUpdated,
            )
        }
        FoodConsumedDetailView(viewModel = viewModel, onBack = onBack)
    }
}
